docs/proposals/job.md

```diff
@@ -0,0 +1,145 @@
+`Job` is responsible for keeping the desired number of Pods to a completion of
+a task. This will be represented by the `RestartPolicy` which is required to
+always take value of `RestartPolicyNever` or `RestartOnFailure`.
+
+
+The new `Job` object will have the following content:
+
+```go
+// Job represents the configuration of a single job.
+type Job struct {
+    TypeMeta
+    ObjectMeta
+
+    // Spec is a structure defining the expected behavior of a job.
+    Spec JobSpec
+
+    // Status is a structure describing current status of a job.
+    Status JobStatus
+}
+
+// JobList is a collection of jobs.
+type JobList struct {
+    TypeMeta
+    ListMeta
+
+    Items []Job
+}
+```
+
+`JobSpec` structure is defined to contain all the information how the actual job execution
+will look like.
+
+```go
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+
+    // TaskCount specifies the desired number of pods the job should be run with.
+    TaskCount int
+
+    // Optional duration in seconds relative to the StartTime that the job may be active
+    // before the system actively tries to terminate it; value must be positive integer
+    ActiveDeadlineSeconds *int64
+
+    // Selector is a label query over pods that should match the pod count.
+    Selector map[string]string
+
+    // Spec is the object that describes the pod that will be created when
+    // executing a job.
+    Spec PodSpec
+}
+```
+
+`JobStatus` structure is defined to contain informations about pods executing
+specified job.  The structure holds information about pods currently executing
+the job.
+
+```go
+// JobStatus represents the current state of a Job.
+type JobStatus struct {
+    // Executions holds a detailed information about each of the pods running a job.
+    Executions []JobExec
+
+    // Completions is the number of pods successfully completed their job.
+    Completions int
+
+}
+
+// JobExec represents the current state of a single execution of a Job.
+type JobExec struct {
+    // CreationTime represents time when the job execution was created
+    CreationTime util.Time
+
+    // StartTime represents time when the job execution was started
+    StartTime util.Time
+
+    // CompletionTime represents time when the job execution was completed
+    CompletionTime util.Time
+
+    // Phase represents the point in the job execution lifecycle.
+    Phase JobExecPhase
+
+    // Tag is added in labels of pod(s) created for this job execution.  It allows
+    // job object to safely group/track all pods started for one given job execution.
+    Tag util.UID
+}
+
+// JobExecPhase represents job execution phase at given point in time.
+type JobExecPhase string
+
+// These are valid JobExec phases.
+const (
+    // JobExecPending means the pod has been accepted by the system but one or more
+    // pods has not been started.
+    JobExecPending JobExecPhase   = "Pending"
```

Comment:
Was JobExecPhase intended to represent the phase of a job or of a task? The comments imply the former whereas the data structure imply the latter.

If the former, and if the semantics are that a number of tasks that may not all run in parallel must be completed, the Pending and Running phases aren’t really appropriate. If the latter, then they seem redundant with pod phases. 

Also, it is not necessary to represent properties that may be inferred from other information. For instance, the JobStatus could contain counts of successful and unsuccessful completions, and the number of currently running tasks. If the RestartPolicy were OnFailure, unsuccessful completions should not be counted, since they would always be retried.

That said, if we wanted to create a polymorphic Runnable interface, we would want to represent in-progress and completion information in a standard, simplified fashion.